runtime/src/iree/hal/drivers/metal/builtin_executables.m

```diff
@@ -6,41 +6,152 @@
 #include "iree/hal/api.h"
 #include "iree/hal/drivers/metal/builtin/metal_buffer_kernels.h"
 
-typedef struct iree_hal_metal_builtin_executable_data_t {
-  const char* entry_point;
+typedef struct iree_hal_metal_builtin_pipeline_info_t {
+  iree_string_view_t entry_point;
   uint32_t file_index;
-} iree_hal_metal_builtin_executable_data_t;
+} iree_hal_metal_builtin_pipeline_info_t;
 
 // The list of builtin executable entry points and their source file index in builtin exectuable
-// embedded data. This MUST be consistent with kernel function names in MSL source code and the file
-// order in embedded data.
-static iree_hal_metal_builtin_executable_data_t iree_hal_metal_builtin_executable_entry_points[] = {
-    {"fill_buffer_16byte", 1},  // Buffer fills; 16-byte aligned offset/length
-    {"fill_buffer_4byte", 1},   // Buffer fills; 4-byte aligned offset/length
-    {"fill_buffer_1byte", 1},   // Buffer fills; 1-byte aligned offset/length
-    {"copy_buffer_1byte", 0},   // Buffer copies; 1-byte aligned offset/length
+// embedded data.
+//
+// NOTE: must be consistent with the same struct in MSL source code.
+// NOTE: the exact order here is assumed below and must not change (that should be fixed...).
+static iree_hal_metal_builtin_pipeline_info_t iree_hal_metal_builtin_pipeline_info[] = {
+    {IREE_SVL("fill_buffer_16byte"), 1},  // Buffer fills; 16-byte aligned offset/length
+    {IREE_SVL("fill_buffer_4byte"), 1},   // Buffer fills; 4-byte aligned offset/length
+    {IREE_SVL("fill_buffer_1byte"), 1},   // Buffer fills; 1-byte aligned offset/length
+    {IREE_SVL("copy_buffer_1byte"), 0},   // Buffer copies; 1-byte aligned offset/length
 };
 
-// The buffer fill specificiation. This MUST be consistent with the same struct in MSL source code.
+// NOTE: must be consistent with the same struct in MSL source code.
 typedef struct iree_hal_metal_buffer_fill_spec_t {
   uint64_t buffer_offset;  // Buffer offset to fill (in bytes)
   uint64_t buffer_length;  // Buffer length to fill (in bytes)
   uint32_t pattern;        // 32-bit fill pattern
 } iree_hal_metal_buffer_fill_spec_t;
 
+// NOTE: must be consistent with the same struct in MSL source code.
 typedef struct iree_hal_metal_buffer_copy_spec_t {
   uint64_t src_buffer_offset;  // Source buffer offset (in bytes)
   uint64_t dst_buffer_offset;  // Destination buffer offset (in bytes)
   uint64_t length;             // Buffer length to fill (in bytes)
 } iree_hal_metal_buffer_copy_spec_t;
 
+// Compiles |source_file| as MSL source into a MTLLibrary for the given |device|.
+//
+// TODO: we should be precompiling this and shipping a binary metallib instead: compiling from
+// source at runtime is _extremely_ inefficient.
+static iree_status_t iree_hal_metal_compile_embedded_msl(id<MTLDevice> device,
+                                                         iree_file_toc_t source_file,
+                                                         id<MTLLibrary>* out_library) {
```

Comment:
Does the compiler need to be running on macOS (with Xcode) to precompile, or is that possible from Linux/Windows hosts? Using spirv-cross or some other tool?

If it requires macOS, we might need to continue supporting both paths, or say that only macOS machines can run the Metal compiler target.

Also, reminder that this doc: https://iree.dev/guides/deployment-configurations/gpu-metal/ is still completely empty >_>

Reply:
the metal tools are available on mac/windows, and I suspect it's runnable on linux under wine - I've also seen https://github.com/darlinghq/darling used to run the mac tools too. something we could do on bots and support as an option for people while still allowing the source fallback.